app/src/main/java/ee/ioc/phon/android/speechutils/AbstractAudioRecorder.java

```diff
@@ -128,20 +152,71 @@ protected int getStatus(int numOfBytes, int len) {
     }
 
     /**
+     * Check if the consume pointer was crossed by the recorded pointer. As long as the consume
+     * pointer was not crossed, the consumption of the buffer may continue as usual and no sound gap
+     * will occur. Once the consume pointer was crossed (e.g. it was on sample 1000 and prior to this
+     * read the recorder was on sample 750 and now that it read the new sample it's on 1500), there's
+     * an audio gap between the consumer and the recorder that can not be filled (data is lost with
+     * no ability to get it back). Whenever this kind of cross occurs, the calling code changed the
+     * session id of the recorder so that if consume is called (from ContinuousRawAudioRecorder),
+     * it will not assume that the data is complete and could be fetched but it will act according to
+     * the SessionStartPointer configured (e.g. read the buffer from the beginning, from now, or from
+     * now - X millis)
+     * @param reachedTheEndOfRecordingBuffer - in case that in the read before the call to this method the recorder
+     *                       passed the end of the buffer and returned to the beginning
+     * @param numOfBytesRead - in the reading process
+     * @return true/false according to the above logic
+     */
+    private boolean isConsumePointerCrossed(boolean reachedTheEndOfRecordingBuffer, int numOfBytesRead) {
```

Comment:
this is a very central piece of code in the sessions mechanism. This is the code that identifies that the recording pointer passed the consuming pointer. In theory this should happen only when we stop consuming (and the consuming pointer stays still while the recording pointer continues). However, for various reasons this case can also happen while consuming (CPU lagging behind due to other activities, lagging due to iops if writing to file etc...)